### PR TITLE
feat(docloaders): add AnswerAgentAI Datastore loaders for Domains, Tags, and Transcripts

### PR DESCRIPTION
**Pull request by bradtaylorsf**
This PR introduces three new LangChain-compatible Document Loaders that pull structured content from the AnswerAgentAI Datastore (Supabase) to power ingestion, chunking, and downstream RAG/search use cases. Each loader exposes rich filters, safe metadata shaping, and an optional text output mode.

What’s included
- AAIDomains loader (packages/components/nodes/documentloaders/AAIDomains/AAIDomains.ts)
  - Filters: searchTerm (domain_name/meta_title/meta_description), includeTags/excludeTags, includeTagsLogic (OR/AND), isValid (all/true/false/null), hasAnalysis (all/analyzed/not_analyzed), limit
  - Pagination with page size up to 100; stops early when results exhausted
  - Reliability: 60s fetch timeout + exponential backoff (3 retries) + small inter-page delay
  - Performance: selects only essential fields; explicitly excludes heavy JSONB history fields
  - Tag handling: flattens domain_tags join to tags array; supports include/exclude logic
  - Content: builds markdown pageContent with AI analysis (optional) + meta info
  - Metadata: includes core fields, AI-derived fields, tag slugs/labels, ISO timestamps, and numeric timestamp variants; source markers (source=aai_datastore, source_type=domain)
  - Metadata shaping: merge additional metadata; omit via keys list or * to keep only custom metadata
  - Output: Document[] or concatenated text (with TextSplitter support)

- AAITags loader (packages/components/nodes/documentloaders/AAITags/AAITags.ts)
  - Filters: searchTerm (slug/label/description), parentTagsOnly, childTagsOnly, limit
  - Optional inclusion of parent tag info via self-join alias
  - Content: simple text summary of tag, description, parent
  - Metadata: core fields, parent refs, hierarchy flags, timestamps, source markers (source=aai_datastore, source_type=tag)
  - Metadata shaping and text output mode consistent with Domains loader

- AAITranscripts loader (packages/components/nodes/documentloaders/AAITranscripts/AAITranscripts.ts)
  - Filters (per visible portion of diff): searchTerm (transcript/summary), dateFrom/dateTo (ISO), includeTags/excludeTags with OR/AND logic, hasAnalysis, sentimentMin/sentimentMax, limit
  - Consistent metadata shaping and output behavior with the other loaders
  - Note: diff snippet was truncated; remaining implementation aligns with the same patterns (Supabase client, filters, Document construction)

- Assets
  - answerai-square-black.png icon added for each loader

Environment variables
- AAI_DATASTORE_SUPABASE_URL
- AAI_DATASTORE_SUPABASE_SERVICE_ROLE_KEY
Notes:
- These use the service role key; do not expose in client/browser contexts.
- Required for all three loaders.

Why
- Provide first-class, configurable ingestion nodes for AAI data (domains, tags, transcripts), enabling targeted datasets via tags, validity/analysis state, and date/sentiment filters.
- Improve resiliency and performance for large domain tables while avoiding heavy JSON fields.

Testing/Verification
- Set AAI_DATASTORE_SUPABASE_URL and AAI_DATASTORE_SUPABASE_SERVICE_ROLE_KEY in your environment
- In the node builder:
  - AAIDomains: try limit=50, includeTags="saas,ecommerce", includeTagsLogic="AND", hasAnalysis="analyzed"; verify tags/metadata and AI analysis sections are present when enabled
  - AAITags: set parentTagsOnly=true then childTagsOnly=true; verify parent info appears when includeParentInfo=true
  - AAITranscripts: filter by dateFrom/dateTo and a couple of tag slugs; adjust sentimentMin/sentimentMax; verify results and metadata
- Switch output to “text” to confirm concatenation and escaping behavior
- Optionally test TextSplitter to ensure documents chunk as expected

Performance/Security
- Domains loader limits selected columns and paginates to reduce payload size
- 60s request timeout, exponential backoff, and inter-page delay help protect the datastore
- Uses service-role key; ensure server-side execution only

Breaking changes
- None

Documentation
- Add “AnswerAgentAI Datastore” section to internal docs covering:
  - Required env vars
  - Node inputs/filters and examples
  - Metadata keys, omit rules, and source markers
  - Known constraints (service-role usage, large-table pagination)

Links
- Diff: https://patch-diff.githubusercontent.com/raw/the-answerai/theanswer/pull/643.diff

Sources:
https://patch-diff.githubusercontent.com/raw/the-answerai/theanswer/pull/643.diff

### CODE REVIEW

**Review by maxtechera — CHANGES_REQUESTED** [review]
# Code Review: feat(docloaders): add AnswerAgentAI Datastore loaders

## Summary
This PR introduces 4 new LangChain-compatible Document Loaders for AAI Datastore integration (Domains, Tags, Transcripts, URLs). The architecture is consistent and includes good resilience features, but there are **critical issues** that must be addressed before merging.

## Critical Issues 🔴

### 1. Service Role Key Exposure Risk
All loaders use `AAI_DATASTORE_SUPABASE_SERVICE_ROLE_KEY` without runtime checks to ensure server-side execution only. If a user accidentally configures a flow that exposes these in a client-accessible way, it could leak the service role key.

**Required:** Add runtime validation that execution is server-side only, or use RLS with more restricted keys.

### 2. Missing Input Validation
No validation for:
- Invalid date formats in AAITranscripts (dateFrom, dateTo)
- Negative or zero limit values
- Invalid sentiment range values (should be 1-10)
- sentimentMin > sentimentMax scenarios
- Invalid includeTagsLogic values (only 'OR' and 'AND' should be accepted)

### 3. HTTP Status Filter Logic Error (AAIUrls)
The status filter creates strings like `'http_status.gte.200,http_status.lt.300'` that won't work correctly with Supabase's `.or()` method. This attempts to create AND conditions within a single clause string, which won't parse correctly.

**Impact:** Status filtering won't work as intended.

### 4. Zero Tests for 2,195 Lines of Code
No tests were added despite adding complex pagination, retry logic, filtering, and error handling across 4 loaders.

## Major Issues 🟡

### 5. Inconsistent Timeout Configuration
Domains and URLs loaders implement 60-second timeouts, but Transcripts and Tags do not. This inconsistency could cause issues.

### 6. Inefficient Pagination with Tag Filtering
Loaders fetch pages of data, then filter by tags in-memory. This causes:
- If filtering is aggressive, the limit may not be reached
- May fetch unnecessary pages if most items are filtered out
- Pagination stops based on fetched count, not filtered count

**Recommendation:** Push tag filtering to the database query.

### 7. Potential Infinite Retry Loop
The retry logic can potentially retry indefinitely if certain conditions aren't met. Add a circuit breaker or max total attempts counter.

## Minor Issues 🔵

### 8. Console.log Statements in Production Code
Multiple `console.log` statements throughout (lines like `console.log('[AAIDomains] Starting load...')`). These should use Flowise's logging infrastructure or be conditional debug logs.

### 9. No Type Definitions for Supabase Schema
The loaders assume specific schema structures but don't include TypeScript types for validation. This makes it harder to catch schema changes.

### 10. Memory Concerns with Large Result Sets
All results are accumulated in memory before processing. For large limits (e.g., 10,000), this could consume significant memory. Consider streaming or document memory requirements.

## What Looks Good ✅

1. **Consistent Architecture** - All 4 loaders follow the same pattern
2. **Resilience Features** - Exponential backoff, retry logic, timeouts (on 2 of them)
3. **Comprehensive Filtering** - Rich filter options for different use cases
4. **Performance Optimization** - Explicitly excludes heavy JSONB fields
5. **Flexible Output Modes** - Document array or text concatenation
6. **Proper LangChain Integration** - Correctly extends BaseDocumentLoader
7. **Source Tracking** - All documents include proper provenance metadata

## Testing Requirements

Before merging, add:
- Unit tests for each loader class
- Integration tests with mocked Supabase responses
- Error handling tests (timeout, retry, network failures)
- Edge cases (empty results, invalid inputs, filtering)
- Pagination logic tests

## Recommendation

**REQUEST CHANGES** - Good architectural foundation but needs critical fixes:

### Must Fix:
1. ✅ Fix HTTP status filter logic in AAIUrls
2. ✅ Add input validation for all parameters
3. ✅ Add security warnings/checks for service role key usage
4. ✅ Add tests (minimum: integration tests with mocked Supabase)
5. ✅ Fix pagination + filtering interaction

### Should Fix:
6. Apply consistent timeout configuration
7. Fix potential infinite retry scenarios
8. Replace console.log with proper logging
9. Document database schema requirements

Once these are addressed, this will be a solid addition to the platform.

File: packages/components/nodes/documentloaders/AAIDomains/AAIDomains.ts [hunk 1]
```diff
@@ -0,0 +1,686 @@
...
+        const supabaseUrl = process.env.AAI_DATASTORE_SUPABASE_URL
+        const supabaseKey = process.env.AAI_DATASTORE_SUPABASE_SERVICE_ROLE_KEY
+
+        if (!supabaseUrl || !supabaseKey) {
+            throw new Error(
+                'AAI Datastore configuration missing. Please set AAI_DATASTORE_SUPABASE_URL and AAI_DATASTORE_SUPABASE_SERVICE_ROLE_KEY environment variables.'
+            )
+        }
+
+        // Convert 'all' to null for RPC function (per API spec)
```
Comment (maxtechera) [inline packages/components/nodes/documentloaders/AAIDomains/AAIDomains.ts:235]: 🔴 **CRITICAL: Service Role Key Security Risk**

These environment variables contain the service role key which has full database access. There's no runtime check to ensure these loaders only execute server-side.

**Risk:** If a user accidentally configures a flow that exposes these in a client-accessible context, the service role key could be leaked.

**Required Actions:**
1. Add runtime validation that execution is server-side
2. Consider using RLS with a more restricted key
3. Add prominent warnings in the node descriptions themselves

**Example check:**
```typescript
if (typeof window !== 'undefined') {
    throw new Error('AAIDomains loader must only be used in server-side contexts')
}
```

File: packages/components/nodes/documentloaders/AAITranscripts/AAITranscripts.ts [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+        const includeTagsLogic = (nodeData.inputs?.includeTagsLogic as string) || 'OR'
+        const excludeTags = nodeData.inputs?.excludeTags as string
+        const hasAnalysis = (nodeData.inputs?.hasAnalysis as string) || 'all'
+        const sentimentMin = nodeData.inputs?.sentimentMin as number
+        const sentimentMax = nodeData.inputs?.sentimentMax as number
+        const metadata = nodeData.inputs?.metadata
+        const _omitMetadataKeys = nodeData.inputs?.omitMetadataKeys as string
+        const output = nodeData.outputs?.output as string
+
+        let omitMetadataKeys: string[] = []
```
Comment (maxtechera) [inline packages/components/nodes/documentloaders/AAITranscripts/AAITranscripts.ts:201]: 🔴 **CRITICAL: Missing Input Validation**

No validation is performed on these sentiment values:
- No check that `sentimentMin <= sentimentMax`
- No check that values are in valid range (1-10)
- No check for negative or invalid values

**Impact:** Invalid inputs could cause incorrect queries or unexpected behavior.

**Fix:**
```typescript
if (sentimentMin !== undefined && sentimentMax !== undefined) {
    if (sentimentMin < 1 || sentimentMin > 10 || sentimentMax < 1 || sentimentMax > 10) {
        throw new Error('Sentiment values must be between 1 and 10')
    }
    if (sentimentMin > sentimentMax) {
        throw new Error('sentimentMin cannot be greater than sentimentMax')
    }
}
```

File: packages/components/nodes/documentloaders/AAIUrls/AAIUrls.ts [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+
+                    // Apply HTTP status filter
+                    if (this.statusFilter.length > 0) {
+                        const statusClauses = this.statusFilter.map((filter) => {
+                            if (filter === 'success') return 'http_status.gte.200,http_status.lt.300'
+                            if (filter === 'redirect') return 'http_status.gte.300,http_status.lt.400'
+                            if (filter === 'client_error') return 'http_status.gte.400,http_status.lt.500'
+                            if (filter === 'server_error') return 'http_status.gte.500'
+                            if (filter === 'informational') return 'http_status.gte.100,http_status.lt.200'
+                            if (filter === 'unknown') return 'http_status.is.null'
```
Comment (maxtechera) [inline packages/components/nodes/documentloaders/AAIUrls/AAIUrls.ts]: 🔴 **CRITICAL: HTTP Status Filter Logic Error**

This filter logic has a fundamental problem. Creating strings like:
```
'http_status.gte.200,http_status.lt.300'
```

Won't work correctly with Supabase's `.or()` method. This attempts to create AND conditions within a single clause string, which Supabase won't parse correctly.

**Impact:** Status filtering won't work as intended. Multiple filters like `['success', 'redirect']` will create incompatible conditions.

**Fix Needed:** Restructure to use proper Supabase query building:
```typescript
// For specific codes
if (specificCodes.length > 0) {
    query = query.in('http_status', specificCodes)
}
// For ranges, need separate queries or use .or() with proper format
```

This needs significant refactoring to work correctly.

File: packages/components/nodes/documentloaders/AAITranscripts/AAITranscripts.ts [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+        }
+
+        // Get environment variables
+        const supabaseUrl = process.env.AAI_DATASTORE_SUPABASE_URL
+        const supabaseKey = process.env.AAI_DATASTORE_SUPABASE_SERVICE_ROLE_KEY
+
+        if (!supabaseUrl || !supabaseKey) {
+            throw new Error(
+                'AAI Datastore configuration missing. Please set AAI_DATASTORE_SUPABASE_URL and AAI_DATASTORE_SUPABASE_SERVICE_ROLE_KEY environment variables.'
+            )
```
Comment (maxtechera) [inline packages/components/nodes/documentloaders/AAITranscripts/AAITranscripts.ts:289]: 🟡 **MAJOR: Inconsistent Timeout Configuration**

AAIDomains and AAIUrls implement a 60-second timeout using AbortController, but AAITranscripts and AAITags don't have this protection.

**Impact:** Transcripts and Tags loaders could hang indefinitely on slow/unresponsive Supabase connections.

**Fix:** Apply the same timeout pattern here:
```typescript
global: {
    fetch: (resource, config) => {
        const controller = new AbortController()
        const timeoutId = setTimeout(() => controller.abort(), 60000)
        return fetch(resource, { ...config, signal: controller.signal })
            .finally(() => clearTimeout(timeoutId))
    }
}
```

File: packages/components/nodes/documentloaders/AAIDomains/AAIDomains.ts [hunk 5]
```diff
@@ -0,0 +1,686 @@
...
+                    const timeoutId = setTimeout(() => controller.abort(), 60000)
+
+                    return fetch(resource, {
+                        ...config,
+                        signal: controller.signal
+                    }).finally(() => clearTimeout(timeoutId))
+                }
+            }
+        })
+
```
Comment (maxtechera) [inline packages/components/nodes/documentloaders/AAIDomains/AAIDomains.ts:362]: 🟡 **MAJOR: Inefficient Pagination with In-Memory Filtering**

The pagination logic fetches pages, then filters by tags in-memory. This causes problems:

**Scenario:**
1. Request limit of 100 domains with aggressive tag filter
2. Fetch 100 domains (page size 100)
3. After tag filtering, only 20 match
4. Pagination stops because `data.length < currentPageSize` is false (100 == 100)
5. User gets only 20 items instead of 100

**Impact:** The requested limit may not be reached if filtering is aggressive.

**Better approach:** Push tag filtering to the database query using Supabase's contains operators, or continue pagination until filtered results meet the limit.

Comment (maxtechera) [inline packages/components/nodes/documentloaders/AAIDomains/AAIDomains.ts:362]: 🔵 **MINOR: Console.log in Production Code**

Console.log statements throughout the loaders should use Flowise's logging infrastructure or be conditional debug logs.

**Suggestion:**
```typescript
if (process.env.DEBUG_LOADERS) {
    console.log('[AAIDomains] Starting load with params:', {...})
}
```

Or better, use a proper logger that Flowise provides.
